<a href="https://colab.research.google.com/github/Shubham-Analyst-coder/Netflix-capstone-project/blob/main/Netflix_project_shubham.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
import plotly.express as px
from datetime import datetime

In [2]:
df=pd.read_csv("/content/netflix_titles.csv")
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [4]:
df.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

In [5]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [6]:
def convert_date(date_str):

    if isinstance(date_str, str):

        date_str = date_str.lstrip()
        try:
            return datetime.strptime(date_str, "%B %d, %Y")
        except ValueError:

            print(f"Warning: Could not convert '{date_str}' to datetime.")
            return None
    else:

        print(f"Warning: Input is not a string: '{date_str}'")
        return None


df['date_added'] = df['date_added'].apply(convert_date)


In [7]:
df_1 = df[df['date_added'].isnull()]
df_1

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories      NaN   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown      NaN   
6795   s6796  TV Show                                      Frasier      NaN   
6806   s6807  TV Show                                      Friends      NaN   
6901   s6902  TV Show                              Gunslinger Girl      NaN   
7196   s7197  TV Show                                     Kikoriki      NaN   
7254   s7255  TV Show                          La Familia P. Luche      NaN   
7406   s7407  TV Show                                        Maron      NaN   
7847   s7848  TV Show                                 Red vs. Blue      NaN   
8182   s8183  TV Show                 The Adventures of Figaro Pho      NaN   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev             NaN   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

     date_added  release_year rating    duration  \
6066        NaT          2013  TV-MA   2 Seasons   
6174        NaT          2018  TV-PG   5 Seasons   
6795        NaT          2003  TV-PG  11 Seasons   
6806        NaT          2003  TV-14  10 Seasons   
6901        NaT          2008  TV-14   2 Seasons   
7196        NaT          2010   TV-Y   2 Seasons   
7254        NaT          2012  TV-14   3 Seasons   
7406        NaT          2016  TV-MA   4 Seasons   
7847        NaT          2015     NR  13 Seasons   
8182        NaT          2015  TV-Y7   2 Seasons   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

In [8]:
df_1.loc[df_1.index[0], 'date_added'] = pd.to_datetime('2012-12-06')
df_1.loc[df_1.index[1], 'date_added'] = pd.to_datetime('2013-04-14')
df_1.loc[df_1.index[2], 'date_added'] = pd.to_datetime('2023-10-12')
df_1.loc[df_1.index[3], 'date_added'] = pd.to_datetime('1994-09-22')
df_1.loc[df_1.index[4], 'date_added'] = pd.to_datetime('2008-01-07')
df_1.loc[df_1.index[5], 'date_added'] = pd.to_datetime('2003-12-22')
df_1.loc[df_1.index[6], 'date_added'] = pd.to_datetime('2002-08-07')
df_1.loc[df_1.index[7], 'date_added'] = pd.to_datetime('2013-05-03')
df_1.loc[df_1.index[8], 'date_added'] = pd.to_datetime('2003-04-01')
df_1.loc[df_1.index[9], 'date_added'] = pd.to_datetime('2012-10-31')


In [9]:
df_1

show_id     type                                        title director  \
6066   s6067  TV Show  A Young Doctor's Notebook and Other Stories      NaN   
6174   s6175  TV Show              Anthony Bourdain: Parts Unknown      NaN   
6795   s6796  TV Show                                      Frasier      NaN   
6806   s6807  TV Show                                      Friends      NaN   
6901   s6902  TV Show                              Gunslinger Girl      NaN   
7196   s7197  TV Show                                     Kikoriki      NaN   
7254   s7255  TV Show                          La Familia P. Luche      NaN   
7406   s7407  TV Show                                        Maron      NaN   
7847   s7848  TV Show                                 Red vs. Blue      NaN   
8182   s8183  TV Show                 The Adventures of Figaro Pho      NaN   

                                                   cast         country  \
6066  Daniel Radcliffe, Jon Hamm, Adam Godley, Chris...  United Kingdom   
6174                                   Anthony Bourdain   United States   
6795  Kelsey Grammer, Jane Leeves, David Hyde Pierce...   United States   
6806  Jennifer Aniston, Courteney Cox, Lisa Kudrow, ...   United States   
6901  Yuuka Nanri, Kanako Mitsuhashi, Eri Sendai, Am...           Japan   
7196                                      Igor Dmitriev             NaN   
7254  Eugenio Derbez, Consuelo Duval, Luis Manuel Áv...   United States   
7406  Marc Maron, Judd Hirsch, Josh Brener, Nora Zeh...   United States   
7847  Burnie Burns, Jason Saldaña, Gustavo Sorola, G...   United States   
8182  Luke Jurevicius, Craig Behenna, Charlotte Haml...       Australia   

     date_added  release_year rating    duration  \
6066 2012-12-06          2013  TV-MA   2 Seasons   
6174 2013-04-14          2018  TV-PG   5 Seasons   
6795 2023-10-12          2003  TV-PG  11 Seasons   
6806 1994-09-22          2003  TV-14  10 Seasons   
6901 2008-01-07          2008  TV-14   2 Seasons   
7196 2003-12-22          2010   TV-Y   2 Seasons   
7254 2002-08-07          2012  TV-14   3 Seasons   
7406 2013-05-03          2016  TV-MA   4 Seasons   
7847 2003-04-01          2015     NR  13 Seasons   
8182 2012-10-31          2015  TV-Y7   2 Seasons   

                                              listed_in  \
6066           British TV Shows, TV Comedies, TV Dramas   
6174                                         Docuseries   
6795                     Classic & Cult TV, TV Comedies   
6806                     Classic & Cult TV, TV Comedies   
6901                       Anime Series, Crime TV Shows   
7196                                           Kids' TV   
7254  International TV Shows, Spanish-Language TV Sh...   
7406                                        TV Comedies   
7847  TV Action & Adventure, TV Comedies, TV Sci-Fi ...   
8182                              Kids' TV, TV Comedies   

                                            description  
6066  Set during the Russian Revolution, this comic ...  
6174  This CNN original series has chef Anthony Bour...  
6795  Frasier Crane is a snooty but lovable Seattle ...  
6806  This hit sitcom follows the merry misadventure...  
6901  On the surface, the Social Welfare Agency appe...  
7196  A wacky rabbit and his gang of animal pals hav...  
7254  This irreverent sitcom featues Ludovico, Feder...  
7406  Marc Maron stars as Marc Maron, who interviews...  
7847  This parody of first-person shooter games, mil...  
8182  Imagine your worst fears, then multiply them: ...

In [10]:
df_1['date_added']=pd.to_datetime(df_1['date_added'])

In [11]:
df=df.set_index("show_id")
df_1=df_1.set_index("show_id")
df.update(df_1)
df.reset_index()

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

     date_added  release_year rating   duration  \
0    2021-09-25          2020  PG-13     90 min   
1    2021-09-24          2021  TV-MA  2 Seasons   
2    2021-09-24          2021  TV-MA   1 Season   
3    2021-09-24          2021  TV-MA   1 Season   
4    2021-09-24          2021  TV-MA  2 Seasons   
...         ...           ...    ...        ...   
8802 2019-11-20          2007      R    158 min   
8803 2019-07-01          2018  TV-Y7  2 Seasons   
8804 2019-11-01          2009      R     88 min   
8805 2020-01-11          2006     PG     88 min   
8806 2019-03-02          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

In [12]:
df.head()

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

        date_added  release_year rating   duration  \
show_id                                              
s1      2021-09-25          2020  PG-13     90 min   
s2      2021-09-24          2021  TV-MA  2 Seasons   
s3      2021-09-24          2021  TV-MA   1 Season   
s4      2021-09-24          2021  TV-MA   1 Season   
s5      2021-09-24          2021  TV-MA  2 Seasons   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   

                                               description  
show_id                                                     
s1       As her father nears the end of his life, filmm...  
s2       After crossing paths at a party, a Cape Town t...  
s3       To protect his family from a powerful drug lor...  
s4       Feuds, flirtations and toilet talk go down amo...  
s5       In a city of coaching centers known to train I...

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8807 entries, s1 to s8807
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   type          8807 non-null   object        
 1   title         8807 non-null   object        
 2   director      6173 non-null   object        
 3   cast          7982 non-null   object        
 4   country       7976 non-null   object        
 5   date_added    8807 non-null   datetime64[ns]
 6   release_year  8807 non-null   int64         
 7   rating        8803 non-null   object        
 8   duration      8804 non-null   object        
 9   listed_in     8807 non-null   object        
 10  description   8807 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(9)
memory usage: 825.7+ KB


In [14]:
df['release_year'] = pd.to_datetime(df['release_year'], format='%Y').dt.year

In [15]:
df

type                  title         director  \
show_id                                                    
s1         Movie   Dick Johnson Is Dead  Kirsten Johnson   
s2       TV Show          Blood & Water              NaN   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans              NaN   
s5       TV Show           Kota Factory              NaN   
...          ...                    ...              ...   
s8803      Movie                 Zodiac    David Fincher   
s8804    TV Show            Zombie Dumb              NaN   
s8805      Movie             Zombieland  Ruben Fleischer   
s8806      Movie                   Zoom     Peter Hewitt   
s8807      Movie                 Zubaan      Mozez Singh   

                                                      cast        country  \
show_id                                                                     
s1                                                     NaN  United States   
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
s4                                                     NaN            NaN   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                    ...            ...   
s8803    Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
s8804                                                  NaN            NaN   
s8805    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
s8806    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
s8807    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

        date_added  release_year rating   duration  \
show_id                                              
s1      2021-09-25          2020  PG-13     90 min   
s2      2021-09-24          2021  TV-MA  2 Seasons   
s3      2021-09-24          2021  TV-MA   1 Season   
s4      2021-09-24          2021  TV-MA   1 Season   
s5      2021-09-24          2021  TV-MA  2 Seasons   
...            ...           ...    ...        ...   
s8803   2019-11-20          2007      R    158 min   
s8804   2019-07-01          2018  TV-Y7  2 Seasons   
s8805   2019-11-01          2009      R     88 min   
s8806   2020-01-11          2006     PG     88 min   
s8807   2019-03-02          2015  TV-14    111 min   

                                                 listed_in  \
show_id                                                      
s1                                           Documentaries   
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   
...                                                    ...   
s8803                       Cult Movies, Dramas, Thrillers   
s8804               Kids' TV, Korean TV Shows, TV Comedies   
s8805                              Comedies, Horror Movies   
s8806                   Children & Family Movies, Comedies   
s8807       Dramas, International Movies, Music & Musicals   

                                               description  
show_id                                                     
s1       As her father nears the end of his life, filmm...  
s2       After crossing paths at a party, a Cape Town t...  
s3       To protect his family from a powerful drug lor...  
s4       Feuds, flirtations and toilet talk go down amo...  
s5       In a city of coaching centers known to train I...  
...                                                    ...  
s8803    A political cartoonist, a crime reporter and a...  
s8804    While living alone in a spooky town, a young g...  
s8805    Looking to survive in a world taken over by zo...  
s8806    Dragged from civilian life, a former su

In [16]:
df.isnull().sum()

type               0
title              0
director        2634
cast             825
country          831
date_added         0
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [17]:
df["director"]=df["director"].fillna("unknown")
df["cast"]=df["cast"].fillna("unknown")
df["country"]=df["country"].fillna("unknown")


In [18]:
df["director"].isnull().sum()

np.int64(0)

In [19]:
df["rating"].isnull().sum()

np.int64(4)

In [20]:
df_1 = df[df['rating'].isnull()]
df_1

type                                              title  \
show_id                                                               
s5990      Movie  13TH: A Conversation with Oprah Winfrey & Ava ...   
s6828    TV Show                  Gargantia on the Verdurous Planet   
s7313    TV Show                                       Little Lunch   
s7538      Movie                               My Honor Was Loyalty   

                director                                               cast  \
show_id                                                                       
s5990            unknown                        Oprah Winfrey, Ava DuVernay   
s6828            unknown  Kaito Ishikawa, Hisako Kanemoto, Ai Kayano, Ka...   
s7313            unknown  Flynn Curry, Olivia Deeble, Madison Lu, Oisín ...   
s7538    Alessandro Pepe  Leone Frisa, Paolo Vaccarino, Francesco Miglio...   

           country date_added  release_year rating  duration  \
show_id                                                        
s5990      unknown 2017-01-26          2017    NaN    37 min   
s6828        Japan 2016-12-01          2013    NaN  1 Season   
s7313    Australia 2018-02-01          2015    NaN  1 Season   
s7538        Italy 2017-03-01          2015    NaN   115 min   

                                    listed_in  \
show_id                                         
s5990                                  Movies   
s6828    Anime Series, International TV Shows   
s7313                   Kids' TV, TV Comedies   
s7538                                  Dramas   

                                               description  
show_id                                                     
s5990    Oprah Winfrey sits down with director Ava DuVe...  
s6828    After falling through a wormhole, a space-dwel...  
s7313    Adopting a child's perspective, this show take...  
s7538    Amid the chaos and horror of World War II, a c...

In [21]:
df.loc["s5990", 'rating'] = ('TV-MA')
df.loc["s6828", 'rating'] = ('TV-14')
df.loc["s7313", 'rating'] = ('G')
df.loc["s7538", 'rating'] = ('R')

In [22]:
df["rating"]

show_id
s1       PG-13
s2       TV-MA
s3       TV-MA
s4       TV-MA
s5       TV-MA
         ...  
s8803        R
s8804    TV-Y7
s8805        R
s8806       PG
s8807    TV-14
Name: rating, Length: 8807, dtype: object

In [23]:
df.update(df_1)

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8807 entries, s1 to s8807
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   type          8807 non-null   object        
 1   title         8807 non-null   object        
 2   director      8807 non-null   object        
 3   cast          8807 non-null   object        
 4   country       8807 non-null   object        
 5   date_added    8807 non-null   datetime64[ns]
 6   release_year  8807 non-null   int64         
 7   rating        8807 non-null   object        
 8   duration      8804 non-null   object        
 9   listed_in     8807 non-null   object        
 10  description   8807 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(9)
memory usage: 1.1+ MB


In [25]:
df[df["duration"].isnull()]

type                                 title    director        cast  \
show_id                                                                        
s5542    Movie                       Louis C.K. 2017  Louis C.K.  Louis C.K.   
s5795    Movie                 Louis C.K.: Hilarious  Louis C.K.  Louis C.K.   
s5814    Movie  Louis C.K.: Live at the Comedy Store  Louis C.K.  Louis C.K.   

               country date_added  release_year  rating duration listed_in  \
show_id                                                                      
s5542    United States 2017-04-04          2017  74 min      NaN    Movies   
s5795    United States 2016-09-16          2010  84 min      NaN    Movies   
s5814    United States 2016-08-15          2015  66 min      NaN    Movies   

                                               description  
show_id                                                     
s5542    Louis C.K. muses on religion, eternal love, gi...  
s5795    Emmy-winning comedy writer Louis C.K. brings h...  
s5814    The comic puts his trademark hilarious/thought...

In [26]:
df.loc["s5542","duration"]="74 min"
df.loc["s5795","duration"]="84 min"
df.loc["s5814","duration"]="66 min"

In [27]:
df.isnull().sum()

type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [28]:
df["type"].unique()

array(['Movie', 'TV Show'], dtype=object)

In [29]:
df['genres'] = df['listed_in'].str.split(',').str[0]

In [30]:
df_1=df[df["type"]=="Movie"]
df_2=df[df["type"]!="Movie"]

In [31]:
df_1

type                             title  \
show_id                                            
s1       Movie              Dick Johnson Is Dead   
s7       Movie  My Little Pony: A New Generation   
s8       Movie                           Sankofa   
s10      Movie                      The Starling   
s13      Movie                      Je Suis Karl   
...        ...                               ...   
s8802    Movie                           Zinzana   
s8803    Movie                            Zodiac   
s8805    Movie                        Zombieland   
s8806    Movie                              Zoom   
s8807    Movie                            Zubaan   

                              director  \
show_id                                  
s1                     Kirsten Johnson   
s7       Robert Cullen, José Luis Ucha   
s8                        Haile Gerima   
s10                     Theodore Melfi   
s13                Christian Schwochow   
...                                ...   
s8802                  Majid Al Ansari   
s8803                    David Fincher   
s8805                  Ruben Fleischer   
s8806                     Peter Hewitt   
s8807                      Mozez Singh   

                                                      cast  \
show_id                                                      
s1                                                 unknown   
s7       Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
s8       Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
s10      Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
s13      Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                    ...   
s8802    Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
s8803    Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
s8805    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
s8806    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
s8807    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                   country date_added  \
show_id                                                                 
s1                                           United States 2021-09-25   
s7                                                 unknown 2021-09-24   
s8       United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
s10                                          United States 2021-09-24   
s13                                Germany, Czech Republic 2021-09-23   
...                                                    ...        ...   
s8802                         United Arab Emirates, Jordan 2016-03-09   
s8803                                        United States 2019-11-20   
s8805                                        United States 2019-11-01   
s8806                                        United States 2020-01-11   
s8807                                                India 2019-03-02   

         release_year rating duration  \
show_id                                 
s1               2020  PG-13   90 min   
s7               2021     PG   91 min   
s8               1993  TV-MA  125 min   
s10              2021  PG-13  104 min   
s13              2021  TV-MA  127 min   
...               ...    ...      ...   
s8802            2015  TV-MA   96 min   
s8803            2007      R  158 min   
s8805            2009      R   88 min   
s8806            2006     PG   88 min   
s8807            2015  TV-14  111 min   

                                                listed_in  \
show_id                                                     
s1                                          Documentaries   
s7                               Children & Family Movies   
s8       Dramas, Independent Movies, International Movies   
s10                                      Comedies, Dramas   
s13                          Dramas, International Movies   
...                                                   ...   
s8802  

In [32]:
df_2

type                  title         director  \
show_id                                                    
s2       TV Show          Blood & Water          unknown   
s3       TV Show              Ganglands  Julien Leclercq   
s4       TV Show  Jailbirds New Orleans          unknown   
s5       TV Show           Kota Factory          unknown   
s6       TV Show          Midnight Mass    Mike Flanagan   
...          ...                    ...              ...   
s8796    TV Show        Yu-Gi-Oh! Arc-V          unknown   
s8797    TV Show             Yunus Emre          unknown   
s8798    TV Show              Zak Storm          unknown   
s8801    TV Show     Zindagi Gulzar Hai          unknown   
s8804    TV Show            Zombie Dumb          unknown   

                                                      cast  \
show_id                                                      
s2       Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
s3       Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...   
s4                                                 unknown   
s5       Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
s6       Kate Siegel, Zach Gilford, Hamish Linklater, H...   
...                                                    ...   
s8796    Mike Liscio, Emily Bauer, Billy Bob Thompson, ...   
s8797    Gökhan Atalay, Payidar Tüfekçioglu, Baran Akbu...   
s8798    Michael Johnston, Jessica Gee-George, Christin...   
s8801    Sanam Saeed, Fawad Khan, Ayesha Omer, Mehreen ...   
s8804                                              unknown   

                                               country date_added  \
show_id                                                             
s2                                        South Africa 2021-09-24   
s3                                             unknown 2021-09-24   
s4                                             unknown 2021-09-24   
s5                                               India 2021-09-24   
s6                                             unknown 2021-09-24   
...                                                ...        ...   
s8796                                    Japan, Canada 2018-05-01   
s8797                                           Turkey 2017-01-17   
s8798    United States, France, South Korea, Indonesia 2018-09-13   
s8801                                         Pakistan 2016-12-15   
s8804                                          unknown 2019-07-01   

         release_year rating   duration  \
show_id                                   
s2               2021  TV-MA  2 Seasons   
s3               2021  TV-MA   1 Season   
s4               2021  TV-MA   1 Season   
s5               2021  TV-MA  2 Seasons   
s6               2021  TV-MA   1 Season   
...               ...    ...        ...   
s8796            2015  TV-Y7  2 Seasons   
s8797            2016  TV-PG  2 Seasons   
s8798            2016  TV-Y7  3 Seasons   
s8801            2012  TV-PG   1 Season   
s8804            2018  TV-Y7  2 Seasons   

                                                 listed_in  \
show_id                                                      
s2         International TV Shows, TV Dramas, TV Mysteries   
s3       Crime TV Shows, International TV Shows, TV Act...   
s4                                  Docuseries, Reality TV   
s5       International TV Shows, Romantic TV Shows, TV ...   
s6                      TV Dramas, TV Horror, TV Mysteries   
...                                                    ...   
s8796                               Anime Series, Kids' TV   
s8797                    International TV Shows, TV Dramas   
s8798                                             Kids' TV   
s8801    International TV Shows, Romantic TV Shows, TV ...   
s8804               Kids' TV, Korean TV Shows, TV Comedies   

                                               description  \
show_id                                                      
s2       After crossing paths at a 

In [33]:
df_1['duration(min)'] = pd.to_numeric(df_1['duration'].str.extract('(\d+)', expand=False), errors='coerce').fillna(0).astype(int)

In [34]:
df_1

type                             title  \
show_id                                            
s1       Movie              Dick Johnson Is Dead   
s7       Movie  My Little Pony: A New Generation   
s8       Movie                           Sankofa   
s10      Movie                      The Starling   
s13      Movie                      Je Suis Karl   
...        ...                               ...   
s8802    Movie                           Zinzana   
s8803    Movie                            Zodiac   
s8805    Movie                        Zombieland   
s8806    Movie                              Zoom   
s8807    Movie                            Zubaan   

                              director  \
show_id                                  
s1                     Kirsten Johnson   
s7       Robert Cullen, José Luis Ucha   
s8                        Haile Gerima   
s10                     Theodore Melfi   
s13                Christian Schwochow   
...                                ...   
s8802                  Majid Al Ansari   
s8803                    David Fincher   
s8805                  Ruben Fleischer   
s8806                     Peter Hewitt   
s8807                      Mozez Singh   

                                                      cast  \
show_id                                                      
s1                                                 unknown   
s7       Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
s8       Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
s10      Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
s13      Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                    ...   
s8802    Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
s8803    Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
s8805    Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
s8806    Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
s8807    Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                   country date_added  \
show_id                                                                 
s1                                           United States 2021-09-25   
s7                                                 unknown 2021-09-24   
s8       United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
s10                                          United States 2021-09-24   
s13                                Germany, Czech Republic 2021-09-23   
...                                                    ...        ...   
s8802                         United Arab Emirates, Jordan 2016-03-09   
s8803                                        United States 2019-11-20   
s8805                                        United States 2019-11-01   
s8806                                        United States 2020-01-11   
s8807                                                India 2019-03-02   

         release_year rating duration  \
show_id                                 
s1               2020  PG-13   90 min   
s7               2021     PG   91 min   
s8               1993  TV-MA  125 min   
s10              2021  PG-13  104 min   
s13              2021  TV-MA  127 min   
...               ...    ...      ...   
s8802            2015  TV-MA   96 min   
s8803            2007      R  158 min   
s8805            2009      R   88 min   
s8806            2006     PG   88 min   
s8807            2015  TV-14  111 min   

                                                listed_in  \
show_id                                                     
s1                                          Documentaries   
s7                               Children & Family Movies   
s8       Dramas, Independent Movies, International Movies   
s10                                      Comedies, Dramas   
s13                          Dramas, International Movies   
...                                                   ...   
s8802  

In [35]:
df_1.describe()

date_added  release_year  duration(min)
count                           6131   6131.000000    6131.000000
mean   2019-05-07 03:32:47.639863040   2013.121514      99.564998
min              2008-01-01 00:00:00   1942.000000       3.000000
25%              2018-04-01 00:00:00   2012.000000      87.000000
50%              2019-06-19 00:00:00   2016.000000      98.000000
75%              2020-07-23 12:00:00   2018.000000     114.000000
max              2021-09-25 00:00:00   2021.000000     312.000000
std                              NaN      9.678169      28.289504

Insigths:

1.Average length of movie runtime is around 99.56 min.

2.(25, 50 & 75) percentile of movie length is 87, 98 & 114 minutes.

3.Standard deviation movie runtime length is 28.28 min.

In [36]:
genres=df["genres"].value_counts()
genres

genres
Dramas                          1600
Comedies                        1210
Action & Adventure               859
Documentaries                    829
International TV Shows           774
Children & Family Movies         605
Crime TV Shows                   399
Kids' TV                         388
Stand-Up Comedy                  334
Horror Movies                    275
British TV Shows                 253
Docuseries                       221
Anime Series                     176
International Movies             128
TV Comedies                      120
Reality TV                       120
Classic Movies                    80
TV Dramas                         67
Thrillers                         65
Movies                            57
TV Action & Adventure             40
Stand-Up Comedy & Talk Shows      34
Romantic TV Shows                 32
Classic & Cult TV                 22
Anime Features                    21
Independent Movies                20
Music & Musicals                  18
TV Shows                          16
Sci-Fi & Fantasy                  13
Cult Movies                       12
TV Horror                         11
Romantic Movies                    3
Spanish-Language TV Shows          2
LGBTQ Movies                       1
TV Sci-Fi & Fantasy                1
Sports Movies                      1
Name: count, dtype: int64

In [37]:
genre_counts = df.groupby('genres').size().reset_index()
genre_counts.rename(columns={0: 'count'}, inplace=True)

In [38]:
fig = px.bar(genre_counts, x='genres', y='count',
             title='Distribution of Content Over Different Genres',
             labels={'genres': 'Genre', 'count': 'Number of Titles'})

In [39]:
fig.show()

In [40]:
df['release_year'].value_counts()

release_year
2018    1147
2017    1032
2019    1030
2020     953
2016     902
        ... 
1961       1
1925       1
1959       1
1966       1
1947       1
Name: count, Length: 74, dtype: int64

In [41]:
release=df.groupby("release_year").size().reset_index()
release

release_year     0
0           1925     1
1           1942     2
2           1943     3
3           1944     3
4           1945     4
..           ...   ...
69          2017  1032
70          2018  1147
71          2019  1030
72          2020   953
73          2021   592

[74 rows x 2 columns]

In [42]:
release.rename(columns={0: 'count'}, inplace=True)

In [43]:
ss=px.bar(release,x="release_year",y="count",title="Release Year")
ss.show()

In [44]:
country=df.groupby("country").size().reset_index()
country

country    0
0                                    , France, Algeria    1
1                                        , South Korea    1
2                                            Argentina   56
3    Argentina, Brazil, France, Poland, Germany, De...    1
4                                     Argentina, Chile    2
..                                                 ...  ...
744                                Venezuela, Colombia    1
745                                            Vietnam    7
746                                       West Germany    1
747                                           Zimbabwe    1
748                                            unknown  831

[749 rows x 2 columns]

In [45]:
country.rename(columns={0: 'count'}, inplace=True)

In [46]:
px.choropleth(country,locations=country['country'],locationmode='country names',projection='orthographic',color = country['count'])

In [47]:
df['date_added'] = pd.to_datetime(df['date_added'])
df['year_added'] = df['date_added'].dt.year
df['month_added'] = df['date_added'].dt.month
monthly_counts = df.groupby(['year_added', 'month_added']).size().reset_index(name='count')
monthly_counts['DAY'] = 1
monthly_counts['Date'] = pd.to_datetime(monthly_counts[['year_added', 'month_added', 'DAY']].rename(columns={'year_added': 'year', 'month_added': 'month', 'DAY': 'day'}))
monthly_counts.set_index('Date', inplace=True)

In [48]:
monthly_counts

year_added  month_added  count  DAY
Date                                           
1994-09-01        1994            9      1    1
2002-08-01        2002            8      1    1
2003-04-01        2003            4      1    1
2003-12-01        2003           12      1    1
2008-01-01        2008            1      2    1
...                ...          ...    ...  ...
2021-06-01        2021            6    207    1
2021-07-01        2021            7    257    1
2021-08-01        2021            8    178    1
2021-09-01        2021            9    183    1
2023-10-01        2023           10      1    1

[116 rows x 4 columns]

In [49]:
fig = px.line(monthly_counts, x=monthly_counts.index, y='count',
             title='Content Added Over Time',
             labels={'Date': 'Date', 'count': 'Number of Titles Added'})

In [50]:
fig.update_layout(xaxis_tickangle=-45)

fig.show()

In [51]:
rating=df.groupby(["rating","type"]).size().reset_index(name="count")
rating

rating     type  count
0     66 min    Movie      1
1     74 min    Movie      1
2     84 min    Movie      1
3          G    Movie     41
4          G  TV Show      1
5      NC-17    Movie      3
6         NR    Movie     75
7         NR  TV Show      5
8         PG    Movie    287
9      PG-13    Movie    490
10         R    Movie    798
11         R  TV Show      2
12     TV-14    Movie   1427
13     TV-14  TV Show    734
14      TV-G    Movie    126
15      TV-G  TV Show     94
16     TV-MA    Movie   2063
17     TV-MA  TV Show   1145
18     TV-PG    Movie    540
19     TV-PG  TV Show    323
20      TV-Y    Movie    131
21      TV-Y  TV Show    176
22     TV-Y7    Movie    139
23     TV-Y7  TV Show    195
24  TV-Y7-FV    Movie      5
25  TV-Y7-FV  TV Show      1
26        UR    Movie      3

In [52]:
fig=px.bar(rating, x="rating",y="count",color='type',title='Distribution of Content Ratings')
fig.show()

In [53]:
df_1["duration(min)"].describe()

count    6131.000000
mean       99.564998
std        28.289504
min         3.000000
25%        87.000000
50%        98.000000
75%       114.000000
max       312.000000
Name: duration(min), dtype: float64

In [54]:
fig=px.histogram(df_1,x="duration(min)",title="Distribution of Movie Length")
fig.show()

In [55]:

jj=px.box(df_1,x="release_year",y="duration(min)",title="Distribution of Movie Length")
jj.show()

.As we can see over here that most of the movie length is between 75 - 137 minutes.

.Chances of Movie length lower or higher than (75 - 137) minutes range is quite low.

.Box Plot also suggest the same. While there might be slight variations across release years, the overall trend suggests consistency in movie lengths

In [56]:
rating=df.groupby("rating").size().reset_index(name="count")
rating

rating  count
0     66 min      1
1     74 min      1
2     84 min      1
3          G     42
4      NC-17      3
5         NR     80
6         PG    287
7      PG-13    490
8          R    800
9      TV-14   2161
10      TV-G    220
11     TV-MA   3208
12     TV-PG    863
13      TV-Y    307
14     TV-Y7    334
15  TV-Y7-FV      6
16        UR      3

In [57]:
fig=px.bar(rating,x="rating",y="count",title="Top Rating",color="rating",template="plotly_dark")
fig.show()

In [58]:
popularity=df.groupby(["release_year","genres"]).size().reset_index(name="count")

In [59]:
fig=px.line(popularity,x="release_year",y="count",color="genres",title="Popularity of Genre")
fig.show()

In [60]:
 country_genre_counts = df.groupby(['country', 'genres']).size().reset_index(name='count')

In [61]:
fig = px.sunburst(country_genre_counts,
                  path=['country', 'genres'],
                  values='count',
                  title='Content Distribution Across Countries and Genres (Sunburst Chart)',
                  labels={'count': 'Number of Titles'})
fig.show()


In [62]:
import pandas as pd

def top3_genres_by_country(df):

    country_genre_counts = df.groupby(['country', 'genres'])['genres'].count().reset_index(name='count')

    country_genre_counts = country_genre_counts.sort_values(['country', 'count'], ascending=[True, False])

    top3_genres = country_genre_counts.groupby('country').head(3)

    return top3_genres

top3_df = top3_genres_by_country(df)
print(top3_df)

                country                    genres  count
0     , France, Algeria                    Dramas      1
1         , South Korea    International TV Shows      1
8             Argentina                    Dramas     14
13            Argentina           Stand-Up Comedy      8
6             Argentina            Crime TV Shows      7
...                 ...                       ...    ...
1409       West Germany             Documentaries      1
1410           Zimbabwe                  Comedies      1
1424            unknown    International TV Shows    159
1415            unknown  Children & Family Movies    105
1421            unknown                    Dramas     72

[1008 rows x 3 columns]


In [63]:
import plotly.express as px

fig = px.bar(top3_df,
             x="country",
             y="count",
             color="genres",
             title="Top 3 Genres by Country",
             labels={"country": "Country", "count": "Number of Titles", "genres": "Genre"},
             barmode='stack')

fig.update_layout(xaxis_tickangle=-100)

fig.show()

In [64]:
a= df.groupby('genres').size().sort_values(ascending=False).reset_index(name = 'count')
a

genres  count
0                         Dramas   1600
1                       Comedies   1210
2             Action & Adventure    859
3                  Documentaries    829
4         International TV Shows    774
5       Children & Family Movies    605
6                 Crime TV Shows    399
7                       Kids' TV    388
8                Stand-Up Comedy    334
9                  Horror Movies    275
10              British TV Shows    253
11                    Docuseries    221
12                  Anime Series    176
13          International Movies    128
14                   TV Comedies    120
15                    Reality TV    120
16                Classic Movies     80
17                     TV Dramas     67
18                     Thrillers     65
19                        Movies     57
20         TV Action & Adventure     40
21  Stand-Up Comedy & Talk Shows     34
22             Romantic TV Shows     32
23             Classic & Cult TV     22
24                Anime Features     21
25            Independent Movies     20
26              Music & Musicals     18
27                      TV Shows     16
28              Sci-Fi & Fantasy     13
29                   Cult Movies     12
30                     TV Horror     11
31               Romantic Movies      3
32     Spanish-Language TV Shows      2
33                  LGBTQ Movies      1
34                 Sports Movies      1
35           TV Sci-Fi & Fantasy      1

In [65]:
px.bar(a,x="genres", y="count",title="Count of different Genres",labels={"genres": "Genres", "count": "Count"})

In [66]:
type_popularity = df.groupby('type').size().reset_index(name='count')
type_popularity = type_popularity.sort_values(by='count', ascending=False)
print(type_popularity)

      type  count
0    Movie   6131
1  TV Show   2676


In [67]:
fig = px.bar(type_popularity, x='type', y='count',
             title='Popularity of Content Types',
             labels={'type': 'Content Type', 'count': 'Number of Titles'},
             color = 'type')
fig.show()

Conclusion:


**1 Content Distribution**: Netflix offers a diverse range of content, with a strong focus on international movies and TV shows. The United States remains a significant contributor, but other countries, particularly India, are emerging as major content creators.



**2 Genre Preferences** : Dramas, comedies, and documentaries are the most common genres available on Netflix, reflecting a broad appeal to different viewer preferences. International movies and TV shows offer a wide variety of genres, catering to diverse tastes.



**3 Content Trends Over Time**: The volume of content added to Netflix has steadily increased over the years, with movies experiencing significant growth in recent years. This demonstrates Netflix's commitment to expanding its content library and catering to evolving viewer demands.



**4 Content Ratings**: While most of the content on Netflix is suitable for mature audiences, a substantial portion is also available for younger viewers, providing a balance between mature and family-friendly options.



**5 Movie Length**: The average length of movies on Netflix is around 99 minutes, with a typical range between 75 to 137 minutes. This consistency in movie length suggests a focus on providing viewers with a manageable viewing experience.



**6 User Preferences**: While direct user rating data is not available, proxy metrics suggest that content with ratings such as "TV-MA" and "TV-14" is more prevalent, possibly indicating a preference for mature and engaging content among Netflix users.



**7 Genre Popularity**: Dramas and comedies are among the most popular genres on Netflix, reflecting a wide appeal to different viewer preferences. International movies and TV shows contribute to a diverse range of genres, catering to specific interests.



**8 Content Diversity**: Netflix showcases a significant diversity of content, with a wide range of genres and categories available. This reflects a commitment to inclusivity and providing a platform for various voices and perspectives.



**9 Content Type Preference**: Movies are slightly more popular than TV shows on Netflix, indicating a user preference for standalone films. However, TV shows remain a significant part of the content library, offering binge-worthy experiences and diverse storytelling formats.

